### PR TITLE
Feature/api reservation model

### PR DESCRIPTION
**Pull request by Lex-23**

### CODE REVIEW

File: coworkings_be/reservation/tests/factory.py [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+
+class ReservationFactory(DjangoModelFactory):
+    class Meta:
+        model = Reservation
+
+    user = factory.SubFactory(UserFactory)
+    working_space = factory.SubFactory(WorkingSpaceFactory)
+    datetime_range = factory.Sequence(
+        (lambda n: DateTimeTZRange(test_dates[n][0], test_dates[n][1]))
```
Comment (andreyPromaster) [inline coworkings_be/reservation/tests/factory.py:26]:
```suggestion
        (lambda n: DateTimeTZRange(datetime.datetime(2022, 11, n, 10, 00, 00), datetime.datetime(2022, 11, n, 10, 00, 00) + datetime.timedelta(hours=8)))
```

File: coworkings_be/reservation/tests/test_api.py [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+        )
+        self.assertEqual(actual_data.working_space.id, expected_data["working_space"])
+        self.assertEqual(self.user.id, expected_data["user"])
+
+    def test_create_reservation_overlaps_time(self):
+        self.create_reservation()
+        token = CustomTokenObtainPairSerializer.get_token(self.user)
+        self.client.force_authenticate(user=self.user, token=token)
+
+        data = {
```
Comment (andreyPromaster) [inline coworkings_be/reservation/tests/test_api.py:206]: use subTest for it

Comment (andreyPromaster) [reply coworkings_be/reservation/tests/test_api.py:206]: You can use just list of dicts not the dict with numbers as keys

File: coworkings_be/reservation/views.py [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+class ReservationListView(ListCreateAPIView):
+    queryset = Reservation.objects.select_related(
+        "working_space", "working_space__type"
+    ).all()
+    serializer_class = ReservationSerializer
+
+    def create(self, request, *args, **kwargs):
+        data = request.data.copy()
+        data["user"] = request.user.id
```
Comment (andreyPromaster) [inline coworkings_be/reservation/views.py:19]:
```suggestion
data = {**request.data, "user": request.user.id}
```

File: coworkings_be/reservation/views.py [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+        data = serializer.validated_data
+        datetime_range = DateTimeTZRange(data["start_time"], data["end_time"])
+
+        try:
+            new_object = Reservation.objects.create(
+                datetime_range=datetime_range,
+                user=data["user"],
+                working_space=data["working_space"],
+            )
+        except IntegrityError:
```
Comment (andreyPromaster) [inline coworkings_be/reservation/views.py:31]: IntegrityError is too common, Could you elaborate on it

File: coworkings_be/reservation/models.py [hunk 5]
```diff
@@ -12,13 +15,39 @@ class Reservation(AuditMixin, models.Model):
 
     user = models.ForeignKey(CustomUser, on_delete=models.CASCADE)
     working_space = models.ForeignKey(WorkingSpace, on_delete=models.CASCADE)
-    start_time = models.DateTimeField()
-    end_time = models.DateTimeField()
-    total_price = PriceField()
+    datetime_range = DateTimeRangeField(default=None)
```
[the remark this answers is not in the document]
Comment (Lex-23) [reply coworkings_be/reservation/models.py:18]: It's miss. I was need this for first approach. Will remove this statement

[the remark this answers is not in the document]
Comment (Lex-23) [reply coworkings_be/reservation/models.py:18]: It's miss. I was need this for first approach. Will remove this statement

File: coworkings_be/coworking/serializers.py [hunk 6]
```diff
@@ -17,7 +17,7 @@ class Meta:
 
 
 class CoworkingSerializer(serializers.ModelSerializer):
-    owners_queryset = CustomUser.objects.all()
+    owners_queryset = CustomUser.objects.filter(role=UserRoles.OWNER)
     owner = serializers.PrimaryKeyRelatedField(queryset=owners_queryset, required=False)
```
[the remark this answers is not in the document]
Comment (Lex-23) [reply coworkings_be/coworking/serializers.py:21]: Ok. But. Are you sure that approach more readable in code?

File: coworkings_be/reservation/tests/factory.py (hunk 1 above)
Comment (Lex-23) [reply coworkings_be/reservation/tests/factory.py:26]: make sense

File: coworkings_be/reservation/tests/test_api.py (hunk 2 above)
Comment (Lex-23) [reply coworkings_be/reservation/tests/test_api.py:206]: got it

File: coworkings_be/reservation/views.py (hunk 3 above)
Comment (Lex-23) [reply coworkings_be/reservation/views.py:19]: got it

File: coworkings_be/reservation/views.py (hunk 4 above)
Comment (Lex-23) [reply coworkings_be/reservation/views.py:31]: I think, I'll make custom Error for this case

File: coworkings_be/reservation/views.py (hunk 3 above)
Comment (Lex-23) [reply coworkings_be/reservation/views.py:19]: No, this approach ruined request. I need: <QueryDict: {'start_time': ['2022-12-02 10:00:00'], 'end_time': ['2022-12-02 18:00:00'], 'working_space': ['1'], 'user': [23]}>
But with yours approach (and I have tried this early too) we get: {'start_time': ['2022-12-02 10:00:00'], 'end_time': ['2022-12-02 18:00:00'], 'working_space': ['1'], 'user': 23}
It is not same

Comment (andreyPromaster) [reply coworkings_be/reservation/views.py:19]: Oh, I completely forgot about it. QueryDict uses deepcopy() in `copy` call. Thats great

File: coworkings_be/reservation/models.py [hunk 7]
```diff
@@ -12,13 +15,39 @@ class Reservation(AuditMixin, models.Model):
...
+
+    @property
+    def end_time(self):
+        return self.datetime_range.upper
+
+    @property
+    def duration(self):
+        return self.end_time - self.start_time
+
+    def total_price(self):
```
Comment (andreyPromaster) [inline coworkings_be/reservation/models.py:50]: It is not good way to count total price, i suppose you can apply discounts and change type of seats

File: coworkings_be/reservation/tests/factory.py (hunk 1 above)
Comment (andreyPromaster) [reply coworkings_be/reservation/tests/factory.py:26]: your previous solution doesn't contain this validation

File: coworkings_be/reservation/models.py (hunk 7 above)
Comment (Lex-23) [reply coworkings_be/reservation/models.py:50]: Sure, I will. This method I created temporary. In future I plan use this method (or other) for count price with discount and other

File: coworkings_be/reservation/tests/factory.py (hunk 1 above)
Comment (Lex-23) [reply coworkings_be/reservation/tests/factory.py:26]: why? 
test_dates = [
    (
        datetime.datetime(2022, 11, n, 10, 00, 00),
        datetime.datetime(2022, 11, n, 10, 00, 00) + datetime.timedelta(hours=8),
    )
    for n in range(1, 31)
]

 `for n in range(1, 31)` - Does it say you something?)

**Review by andreyPromaster — APPROVED** [review]

File: coworkings_be/coworking/serializers.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (andreyPromaster) [reply coworkings_be/coworking/serializers.py]: I suppose you can guess what's going on in this code without naming  `owners_queryset` :)

**Pull request merged**